src/Tools/dotnet-monitor/ConfigurationBuilderExtensions.cs

```diff
@@ -0,0 +1,94 @@
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Configuration.KeyPerFile;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.Primitives;
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection;
+using System.Threading;
+
+namespace Microsoft.Diagnostics.Tools.Monitor
+{
+    internal static class ConfigurationBuilderExtensions
+    {
+        // Works arounds an isssue in KeyPerFileConfigurationProvider where the change token is not signaled
+        // when the keys and values are reloaded. See https://github.com/dotnet/aspnetcore/issues/32836
+        public static IConfigurationBuilder AddKeyPerFileWithChangeTokenSupport(this IConfigurationBuilder builder, string directoryPath, bool optional, bool reloadOnChange)
+        {
+            return builder.Add(delegate (KeyPerFileConfigurationSourceWithChangeTokenSupport source)
+            {
+                if (!optional || Directory.Exists(directoryPath))
+                {
+                    source.FileProvider = new PhysicalFileProvider(directoryPath);
+                }
+                source.Optional = optional;
+                source.ReloadOnChange = reloadOnChange;
+            });
+        }
+
+        private class KeyPerFileConfigurationSourceWithChangeTokenSupport :
+            KeyPerFileConfigurationSource,
+            IConfigurationSource
+        {
+            IConfigurationProvider IConfigurationSource.Build(IConfigurationBuilder builder)
+            {
+                return new KeyPerFileConfigurationProviderWithChangeTokenSupport(this);
+            }
+        }
+
+        private class KeyPerFileConfigurationProviderWithChangeTokenSupport :
+            KeyPerFileConfigurationProvider,
+            IDisposable
+        {
+            // KeyPerFileConfigurationProvider.Load(bool) method
+            private readonly static MethodInfo s_loadMethod = typeof(KeyPerFileConfigurationProvider).GetMethod(
+                "Load",
+                BindingFlags.NonPublic | BindingFlags.Instance,
+                null,
+                new[] { typeof(bool) },
+                null);
+
+            private readonly IDisposable _changeTokenRegistration;
+
+            public KeyPerFileConfigurationProviderWithChangeTokenSupport(KeyPerFileConfigurationSourceWithChangeTokenSupport source)
+                : base(source)
+            {
+                Debug.Assert(null != s_loadMethod);
+                if (null != s_loadMethod)
+                {
+                    if (source.ReloadOnChange && source.FileProvider != null)
+                    {
+                        _changeTokenRegistration = ChangeToken.OnChange(
+                            () => source.FileProvider.Watch("*"),
+                            () =>
+                            {
+                                Thread.Sleep(source.ReloadDelay);
```

Comment:
I take it there is no async version of ChangeToken.OnChange? Thread.sleep is probably not ideal here, although it shouldn't happen that often.

Reply:
I'm not sure, but I'm guessing there isn't. This is the code that the underlying KeyPerFileConfigurationProvider uses for reload. I just replicated it here in order to plumb in the OnReload call.

Reply:
Checked this and there is no async version.